torch_xla/experimental/scan.py

```diff
@@ -82,14 +88,335 @@ def scan(fn, init, xs):
+  # - If yes, find the position in the fake tensor list. Index into the input ops.
+  # - If no, find the op in the additional input list.
+  def call_fn_computation(carry: List[xb.Op], x: List[xb.Op],
+                          additional_inputs: Tuple[xb.Op, ...]) -> xb.Op:
+    param_id_to_fake_tensors_id: Dict[int, int] = {}
+    for i, v in enumerate(itertools.chain(fake_carry, fake_x)):
+      param_id = fn_ctx.tensor_parameter_id(v)
+      if param_id != -1:
+        param_id_to_fake_tensors_id[param_id] = i
+
+    all_inputs = carry + x
+    all_inputs_reordered = []
+    mapping: Dict[int, torch.Tensor] = fn_ctx.parameter_id_tensor_mapping()
+    for i in range(len(mapping)):
+      if i in param_id_to_fake_tensors_id:
+        op = all_inputs[param_id_to_fake_tensors_id[i]]
+        all_inputs_reordered.append(op)
+      else:
+        op = additional_inputs[param_id_to_additional_tensors_param_id[i] -
+                               len(loop_tensors)]
+        all_inputs_reordered.append(op)
+    return xb.Op.call(fn_computation, all_inputs_reordered)
+
+  @skip_additional_inputs
+  def cond_fn(num_iters: xb.Op, *args: xb.Op):
+    return num_iters > xb.Op.scalar(num_iters.builder(), 0, dtype=xb.Type.S64)
+
+  @pass_through_additional_inputs
+  def body_fn(num_iters: xb.Op, *args: xb.Op, additional_inputs: Tuple[xb.Op,
+                                                                       ...]):
+    carry, xs, fn_carry_history, ys = unpack(args)
+    xs_len_op = xb.Op.scalar(num_iters.builder(), xs_len, dtype=xb.Type.S64)
+    one = xb.Op.scalar(num_iters.builder(), 1, dtype=xb.Type.S64)
+    idx = xs_len_op - num_iters
+    x = [dynamic_slice(v, idx) for v in xs]
+    for i in range(len(carry)):
+      fn_carry_history[i] = dynamic_update_slice(fn_carry_history[i], carry[i],
+                                                 idx)
+    result = call_fn_computation(carry, x, additional_inputs)
+    for i in range(flat_carry_len):
+      carry[i] = result.get_tuple_element(i)
+    for i in range(flat_y_len):
+      y = result.get_tuple_element(i + flat_carry_len)
+      ys[i] = dynamic_update_slice(ys[i], y, idx)
+    return (num_iters - one,) + pack(carry, xs, fn_carry_history, ys)
+
+  res = xb.Op.mkwhile(builder.params(), cond_fn, body_fn)
+  computation = res.build('scan')
+
+  outputs = torch_xla._XLAC._xla_user_computation('xla::scan',
+                                                  builder.param_tensors(),
+                                                  computation)
+  # skip the last num_additional_inputs
+  outputs = outputs[:len(outputs) - num_additional_inputs]
+  # `1:` to skip `num_iters`
+  carry, xs, fn_carry_history, ys = unpack(outputs[1:])
+
+  # Unflatten tensors back to PyTrees
+  return tree_unflatten(carry, carry_spec), tree_unflatten(
+      fn_carry_history, carry_spec), tree_unflatten(ys, fn_output_y_spec)
+
+
+@pytreeify
+class Scan(torch.autograd.Function):
```

Comment:
Note to self: enforce checkpointing by adding the right optimization barriers, and document this constraint. This may be improved in the future. This is fine since checkpointing the decoder block in LLMs is a common use case.

Reply:
I believe the barriers won't be needed. This is the recommendation from JAX, which says you don't want to wrap inputs into a barrier if the checkpointed function is to be used in a `scan`.

[1]: https://jax.readthedocs.io/en/latest/_autosummary/jax.checkpoint.html